apis/python/src/tiledbsoma/pytiledbsoma.cc

```diff
@@ -693,5 +694,69 @@ PYBIND11_MODULE(pytiledbsoma, m) {
         .def("get_enum_is_ordered", get_enum_is_ordered)
 
         .def("get_enum_label_on_attr", &SOMAArray::get_enum_label_on_attr);
+    // Efficient C++ re-indexing (aka hashing unique key values to an index
+    // between 0 and number of keys - 1) based on khash
+    py::class_<IntIndexer>(m, "IntIndexer")
+        .def(py::init<>())
+        .def(py::init([](py::array_t<int64_t> keys, int num_threads) {
+            auto buffer = keys.request();
+            int64_t* data = static_cast<int64_t*>(buffer.ptr);
+            size_t length = buffer.shape[0];
+            IntIndexer indexer(data, length, num_threads);
+            return indexer;
+        }))
+        .def(
+            "map_locations",
+            [](IntIndexer& indexer,
+               py::array_t<int64_t> keys,
+               int num_threads) {
+                auto buffer = keys.request();
+                int64_t* data = static_cast<int64_t*>(buffer.ptr);
+                size_t length = buffer.shape[0];
+                indexer.map_locations(keys.data(), keys.size(), num_threads);
+            })
+        .def(
+            "map_locations",
+            [](IntIndexer& indexer,
+               std::vector<int64_t> keys,
+               int num_threads) {
+                indexer.map_locations(keys.data(), keys.size(), num_threads);
+            })
```

Comment:
It looks like this is only ever going to be used like

```python
indexer = clib.IntIndexer()
indexer.map_locations(...)
# pass indexer off to be used elsewhere
```

In this case, I would just make it so you pass the relevant data directly into the initializer, so there’s no way to end up with an invalid data structure:

```python
indexer = clib.IntIndexer(...)
# pass it off to be used elsewhere
```

Reply:
`pybind11` imports constructors from C++ directly. For example here I can use my map_locations inside the API's map_location but that's not possible for constructors. For example, in this case, for the non-copy constructor to work it must have `IntIndexer::IntIndexer(py::array_t<int64_t>& keys)`, int but the non-copy constructor of my C++ IntIndexer has the following signature `IntIndexer::IntIndexer(const int64_t* keys, int size, int threads)`. Opening up `py::*` types to C++ is something that I strongly tried to avoid.